python-frontend/src/main/java/org/sonar/python/types/v2/TypeCheckBuilder.java

```diff
@@ -269,6 +274,27 @@ public TriBool test(PythonType pythonType) {
     }
   }
 
+  record IsGenericPredicate() implements TypePredicate {
+
+    @Override
+    public TriBool test(PythonType pythonType) {
+      return isGeneric(pythonType);
+    }
+
+    private static TriBool isGeneric(PythonType pythonType) {
+      if (pythonType instanceof ClassType classType && classType.isGeneric()) {
+        return TriBool.TRUE;
+      }
+      if (pythonType instanceof SpecialFormType specialFormType) {
+        return "typing.Generic".equals(specialFormType.fullyQualifiedName()) ? TriBool.TRUE : TriBool.UNKNOWN;
+      }
+      if (pythonType instanceof UnknownType.UnresolvedImportType unresolvedImportType && "typing.Generic".equals(unresolvedImportType.importPath())) {
+        return TriBool.TRUE;
+      }
+      return TriBool.UNKNOWN;
```

Comment:
this function never returns false. Is it a conservative choice? For example, It seems paradoxical for `ClassType` to have `isGeneric` false and to return `UNKNOWN` here.

Reply:
Yes, it's conservative. I think that `isGeneric` in `ClassType` should probably be a `TriBool` in fact, as what I want to express is that `isGeneric=true` means that it's for sure a generic class, and `false` just means there's no sign that it is.
I don't think it should have an impact until we actually want to raise issues on explicitly non generic classes, where this heuristic will have to be thought about more deeply.